File: app/src/main/java/com/numero/sojodia/Activity/MainActivity.java
package com.numero.sojodia.Activity;

import android.content.Intent;
import android.content.res.Configuration;
import android.os.Bundle;
import android.support.design.widget.TabLayout;
import android.support.v4.view.ViewPager;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.View;

import com.numero.sojodia.Adapter.BusScheduleFragmentPagerAdapter;
import com.numero.sojodia.View.UpdateNotificationDialog;
import com.numero.sojodia.Network.UpdateChecker;
import com.numero.sojodia.R;
import com.numero.sojodia.Util.PreferenceUtil;
import com.numero.sojodia.Util.Constants;
import com.numero.sojodia.Util.DateUtil;
import com.numero.sojodia.Util.NetworkUtil;

public class MainActivity extends AppCompatActivity {

    private long versionCode = 0L;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        Configuration configuration = getResources().getConfiguration();
        switch (configuration.orientation) {
            case Configuration.ORIENTATION_PORTRAIT:
                toolbar.setVisibility(View.VISIBLE);
                break;
            case Configuration.ORIENTATION_LANDSCAPE:
                toolbar.setVisibility(View.GONE);
                break;
        }

        BusScheduleFragmentPagerAdapter fragmentPagerAdapter = new BusScheduleFragmentPagerAdapter(this, getSupportFragmentManager());
        ViewPager viewPager = (ViewPager) findViewById(R.id.container);
        viewPager.setAdapter(fragmentPagerAdapter);
        TabLayout tabLayout = (TabLayout) findViewById(R.id.tabs);
        tabLayout.setupWithViewPager(viewPager);

        if (NetworkUtil.canNetworkConnect(this)) {
            if (isTodayCheckUpdate()) {
                return;
            }

            UpdateChecker checker = UpdateChecker.init().setCallback(new UpdateChecker.Callback() {
                @Override
                public void onSuccess(long versionCode) {
                    PreferenceUtil.setUpdatedDate(MainActivity.this, DateUtil.getTodayStringOnlyFigure());
                    if (canUpdate(versionCode)) {
                        MainActivity.this.versionCode = versionCode;
                        showUpdateNotificationDialog();
                    }
                }

                @Override
                public void onFailure() {
                }
            });
            checker.execute();
        }
    }

    private void showUpdateNotificationDialog() {
        UpdateNotificationDialog dialog = new UpdateNotificationDialog(this) {
            @Override
            public void onClickPositiveButton() {
                updateExecute();
            }
        };
        dialog.show();
    }

    private void updateExecute() {
        Intent intent = new Intent(this, UpdateActivity.class);
        intent.putExtra(Constants.VERSION_CODE, versionCode);
        startActivity(intent);
        finish();
    }

    private boolean isTodayCheckUpdate() {
        if (PreferenceUtil.getPreviousUpdatedDate(this).equals("")) {
            return false;
        } else if (!PreferenceUtil.getPreviousUpdatedDate(this).equals(DateUtil.getTodayStringOnlyFigure())) {
            return false;
        }
        return true;
    }

    private boolean canUpdate(long versionCode) {
        return PreferenceUtil.getVersionCode(this) < versionCode;
    }
}

File: app/src/main/java/com/numero/sojodia/activity/MainActivity.kt
package com.numero.sojodia.activity

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.Bundle
import android.support.v4.content.LocalBroadcastManager
import android.support.v7.app.AppCompatActivity
import android.view.Menu
import android.view.MenuItem

import com.numero.sojodia.fragment.BusScheduleFragment
import com.numero.sojodia.model.Reciprocate
import com.numero.sojodia.model.Route
import com.numero.sojodia.presenter.BusSchedulePresenter
import com.numero.sojodia.repository.BusDataRepository
import com.numero.sojodia.view.TimeTableDialog
import com.numero.sojodia.view.adapter.BusScheduleFragmentPagerAdapter
import com.numero.sojodia.service.UpdateBusDataService
import com.numero.sojodia.util.BroadCastUtil
import com.numero.sojodia.view.NeedRestartDialog
import com.numero.sojodia.R
import com.numero.sojodia.extension.getApplicationComponent
import com.numero.sojodia.extension.getTodayString
import kotlinx.android.synthetic.main.activity_main.*
import java.util.*

import javax.inject.Inject

class MainActivity : AppCompatActivity(), BusScheduleFragment.BusScheduleFragmentListener {

    private val finishDownloadReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            showNeedRestartDialog()
        }
    }

    private val changedDateReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            toolbar.subtitle = Calendar.getInstance().getTodayString(getString(R.string.date_pattern))
        }
    }

    @Inject
    lateinit var busDataRepository: BusDataRepository

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        setSupportActionBar(toolbar)
        getApplicationComponent()?.inject(this)

        initViews()

        val shortcutIntent = intent.getStringExtra("shortcut")
        viewPager.currentItem = if (shortcutIntent == "coming_home") 1 else 0
        startCheckUpdateService()
    }

    override fun onResume() {
        super.onResume()
        toolbar.subtitle = Calendar.getInstance().getTodayString(getString(R.string.date_pattern))
    }

    override fun onStart() {
        super.onStart()
        LocalBroadcastManager.getInstance(this).apply {
            registerReceiver(finishDownloadReceiver, IntentFilter(BroadCastUtil.ACTION_FINISH_DOWNLOAD))
            registerReceiver(changedDateReceiver, IntentFilter(BroadCastUtil.ACTION_CHANGED_DATE))
        }
    }

    public override fun onStop() {
        super.onStop()
        LocalBroadcastManager.getInstance(this).apply {
            unregisterReceiver(finishDownloadReceiver)
            unregisterReceiver(changedDateReceiver)
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_settings -> startActivity(SettingsActivity.createIntent(this))
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onActivityCreated(fragment: BusScheduleFragment, reciprocate: Reciprocate) {
        // FIXME AdapterではActivity再生成時にPresenterを生成できないからこの実装している
        BusSchedulePresenter(fragment, busDataRepository, reciprocate)
    }

    override fun showTimeTableDialog(route: Route, reciprocate: Reciprocate) {
        TimeTableDialog(this, busDataRepository).apply {
            setRoute(route)
            setReciprocate(reciprocate)
            show()
        }
    }

    private fun initViews() {
        val fragmentPagerAdapter = BusScheduleFragmentPagerAdapter(this, supportFragmentManager)
        viewPager.adapter = fragmentPagerAdapter
        tabLayout.setupWithViewPager(viewPager)
    }

    private fun startCheckUpdateService() {
        startService(Intent(this@MainActivity, UpdateBusDataService::class.java))
    }

    private fun showNeedRestartDialog() {
        NeedRestartDialog.init(this).setOnPositiveButtonClickListener {
            busDataRepository.clearCache()
            this.recreate()
        }.show()
    }
}
